MO-417 oauth2 authentication

starswan opened the pull request:
This PR creates an auth layer for the existing read endpoint, that just checks that the provided token is valid and has not expired.

starswan committed:
MO-417 working validation of issuer and dynamic public key from HMPPS Auth

ollietreend commented on app/controllers/application_controller.rb:
```diff
@@ -1,4 +1,29 @@
+private
+
+  def authenticate!
+    # OAuth2::Client.new("client_id", "client_secret",
+    #                    :site => "http://localhost:3000")
```
Looks like this comment isn't needed any more

ollietreend commented on deploy/staging/shared-environment.yaml:
```diff
@@ -6,4 +6,5 @@ data:
   RAILS_ENV: "production"
   RAILS_WEB_CONCURRENCY: "5"
   SENTRY_CURRENT_ENV: "staging"
+  NOMIS_OAUTH_PUBLIC_KEY: "LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0KTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FROEFNSUlCQ2dLQ0FRRUF6Y3g3WWJ3MkJBV3Y3U3NFSVVHUAppeUpTSURndHFCeDE5VHdtN3VJM1RYNXpSY3JabFV4Y0VtUEpHVWd5K0QySkloVmxxbWVxd0dWMkNOT3FaQmdHCmo4ZUpHQTU5aUlUemU4ZG1SSk5JYzdsNmxESmc5RE5KVk9pTHFVbFpGRENJcXplSTYzb3E2dWhjY2c1RFBpVE4KcU9HWmM4dXBOK3c1ZFpyTnYrMkdMZ3hLMnBldE1VL0JoWWVXZjNLdllJTzE2djF1dm5GT0dPMTNI ...
```
Is it safe to remove this environment variable now that we're pulling it in dynamically?

ollietreend committed:
Add oAuth 2 security to Swagger definition

ollietreend committed:
Cache the HMPPS Auth JWKS to avoid repeated calls

The HMPPS Auth JWKS is now cached for 24 hours to avoid us getting it
from the server with every request. It doesn't change frequently, so
once a day should be fine for now.

RSpec tests don't have caching enabled, so they will continue to call
out to the server repeatedly. This isn't necessarily a bad thing, as it
means we end up testing the integration between us and HMPPS Auth.

The dev environment doesn't have caching enabled by default. This can be
toggled on/off by running `rails dev:cache`.
